analytics/core.go

```diff
@@ -93,3 +96,8 @@ type NotificationEvent struct {
 	Request *EventRequest   `json:"request"`
 	Account *config.Account `json:"account"`
 }
+
+// LogObject is common loggable object
+type LogObject struct {
+	SeatNonBid []openrtb_ext.SeatNonBid
+}
```

Comment:
`LogObject` is introduced because of following use case
1. Exchange (exchange/exchange.go) suppose to update seat non-bid scenarios to Prebid Analytics Module
2. As per the current design (master) `exchange/exchange.go` has no access to `AuctionObject`, `AmpObject` or `VideoObject`. So, it will not be able to update unless these objects made available to it
3. Today,  `AuctionObject`, `AmpObject` or `VideoObject` are constructed by `auction.go`, `amp_auction.go` and `video_auction.go` respectively
4.  So, I have passed them over `AuctionRequest`. For this I have introduced `LogObject`, the more generalised form Of `AuctionObject`, `AmpObject` or `VideoObject here` [here](https://github.com/PubMatic-OpenWrap/prebid-server/blob/042398562aad7680056520a5ed6243a8062bd24e/exchange/exchange.go#L192) - . Otherwise, we will need to keep 3 different objects within `AuctionRequest`, which I think bad design.

Reply:
I see, this makes it more clear as to why `LogObject` could be valuable, but it also leaves me wondering if we still could just have `SeatNonBid` live in all of the places that `LogObject` currently occupies. I'm going to speak with @SyntaxNode about it, and get back to you!

Reply:
@AlexBVolcy : Have you got a chance to check around it with @SyntaxNode ?

Reply:
Yes. We reviewed the approach together and I understand your intention with introducing the LogObject. Thank you for sharing your use case, that was very useful context.

I agree with your second point that the exchange package has no access to the analytics package. This proposal changes that by adding a direct reference to analytics.LogObject. The exchange package (and HoldAuction in particular) is a weak point of PBS' design as it does too much work and takes too many parameters. The AuctionRequest struct was created as an attempt to curb the growth of HoldAuction, but even that has grow beyond comfort in the last year. For this reason, I ask you to consider a different approach to avoid entangling HoldAuction with more parts of the code.

Instead of passing the log object to HoldAuction, how about we enhance the return of HoldAuction to include the non bid information and handle sharing that with the analytics package in each endpoint? This would align well with future refactor plans, such that each endpoint constructs its own response for its own format. The AMP endpoint does extra work to form it's response which we can simplify in a future refactor, and sending back parts of the response as a strongly typed object would be both a great first step and avoid adding a dependency to the exchange.

What do you think?

Reply:
Yes, I think, it would be good to return of HoldAuction to include the non bid information

Reply:
Thanks to @viral for pointing this out. 
@SyntaxNode : 

**Please provide your views on inteoducing `ResponseWrapper`**

As per the thought, he have we can introduced `ResponseWrapper`, wherein `openrtb2.BidResponse` will be embedded. `ResponseWrapper` will contain the unmarshalled version of `response extension`. This will help us in sharing the `SeatNonBid` object to analytics, which has build by the exchange it self. Structure will look like as follows
```go
type ResponseWrapper struct {
	openrtb2.BidResponse
	responseExt *openrtb_ext.ExtBidResponse
}


func (rw *ResponseWrapper) GetSeatNonBid() []SeatNonBid {
	// return seatnonbid using responseExt
}

// Change HoldAuction signature to return ResponseWrapper instead of openrtb2.BidResponse
HoldAuction(ctx context.Context, r AuctionRequest, debugLog *DebugLog) (*openrb_ext.ResponseWrapper, error)
```
**Benefits of this approach**
1. No need to unmarshal again for obtaining seatnonbid
2. SeatNonBid will be shared, so no new object will be allocated on stack/heap
3. ResponseWrapper is extensible. 


**ReturnAllBidStatus** @SyntaxNode 
1. As per current PR's implementation, exchange.makeExtBidResponse is setting `SeatNonBid` within prebid extension only if   `ReturnAllBidStatus` is true.
2. With current approach, I strongly feel that this needs to move at `auction.go`, `amp_auction.go` and `video_auction.go` level. Otherwise, SeatNonBid can not be made available to Analytics if `ReturnAllBidStatus` is `false`

Reply:
I like your thought process. I am most intrigued with your last arguments pointing to a "code smell" - a design flaw with the current implementation. The auction, amp, and video endpoints each have their own response structure but HoldAuction is built just for auction (the first endpoint we built). We use a hack for the amp and video endpoints where a lot of work is necessary to reformat the json response.

While the ResponseWrapper is a great idea and may be part of the final design, it doesn't fully solve the problem - as you realize with the mention of the `ReturnAllBidStatus` issue.

I agree with your ideal solution of moving exchange.makeExtBidResponse out of HoldAuction and into the auction, amp, and video endpoints. Instead of returning the bid response data structure, HoldAuction can return the parts needed to build the response, including all seat non bid data.

Reply:
Thanks @SyntaxNode  for your review,  I will do the needful and update here